_drafts/tracking-code-size.md

```diff
@@ -0,0 +1,598 @@
+{:toc}
+
+## Why Track Code Size?
+
+If you are a firmware engineer, you've likely run into code sizes issues. If
+not, it is likely because the firmware engineers before you fought for an MCU
+with more micro-flash space or because the company went overboard and bought the
+most expensive chip.
+
+A team that has extra space in the firmware image will find a way to fill it,
+whether it is extra debug logs, a full-featured command-line interface, or fancy
+animations or images. With this in mind, it would be wise to constantly track
+and measure the code size of a project during development to know when and why
+the code size grows. This would mean that each commit and a handful of build
+configurations should be tracked so an accurate picture can be painted.
+
+I've seen many projects start to have a print out like below after each build,
+which is nice, but it isn't enough to keep the code size from being gobbled up.
+
+```
+Memory region         Used Size  Region Size  %age Used
+           FLASH:      153108 B         1 MB     14.60%
+            SRAM:       44544 B       256 KB     16.99%
+        IDT_LIST:          56 B         2 KB      2.73%
+```
+
+Once you have the code size for each commit and build combination, you and your
+entire team can:
+
+- Prevent large (possibly accidental) code size increases due to non-optimal
+  code for function usage (e.g. a few years ago, by adding a call to newlib's
+  `sscanf(...)`, I increased the firmware size by 2kB).
+- You can retroactively find large spikes or decreases in code size when looking
+  for easy-win optimizations.
+- Estimate when the code size will exceed the limit by looking at historical
+  data.
+- Over time, learn which coding patterns contribute to the code size increases.
+- Calculate code size deltas on every new pull request. You would likely prefer
+  to merge a shiny new feature which adds 2kB of code instead of a poorly
+  implemented bug-fix which also adds 2kB.
+
+## How to Track Code Size
+
+In the remainder of the blog post, we'll go through an example of tracking the
+code size of the popular open-source RTOS, Zephyr[^zephyr]. We also want to take
+this data and be able to compute the code size **deltas** between commits to
+know exactly which commits and by how much each one contributed to the code
+size.
+
+Since the Zephyr project is massive in the number of examples, boards, and
+features that it encompasses, we will restrict the code size calculations to the
+`examples/net/lwm2m_client` example with the `cc3220sf_launchxl` board. This
+example was chosen due to the large number of files and dependencies with a hope
+that its code size would fluctuate more than a hello world example with limited
+dependencies.
+
+We won't be covering how to initially set up building the Zephyr project, as
+there are detailed instructions[^zephyr_instructions] about how to get started.
+I can't help but feel the setup would have been simpler and quicker if they had
```

Comment:
"If anyone working on Zephyr is reading this, you really how to check out Conda to make the setup easier ;-)"